core/src/main/java/io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java

```diff
@@ -77,22 +85,58 @@ public long size() {
         return -1;
     }
 
-    protected long getBaseRecordTimestamp() {
-        return timestampSampler.round(baseRecord.getTimestamp(timestampIndex) - baselineOffset);
-    }
-
-    @Override
-    public void of(RecordCursor base, SqlExecutionContext executionContext) {
+    public void of(
+            RecordCursor base,
+            SqlExecutionContext executionContext,
+            Function timezoneNameFunc,
+            int timezoneNameFuncPos,
+            Function offsetFunc,
+            int offsetFuncPos
+    ) throws SqlException {
         // factory guarantees that base cursor is not empty
+        timezoneNameFunc.init(base, executionContext);
+        offsetFunc.init(base, executionContext);
+
+        long alignmentOffset = Numbers.LONG_NaN;
+        final CharSequence tz = timezoneNameFunc.getStr(null);
+        if (tz != null) {
+            try {
+                alignmentOffset = Timestamps.toTimezone(0, TimestampFormatUtils.enLocale, tz);
+            } catch (NumericException e) {
+                Misc.free(base);
+                throw SqlException.$(timezoneNameFuncPos, "invalid timezone: ").put(tz);
+            }
+        }
+
+        CharSequence offset = offsetFunc.getStr(null);
+
+        if (offset != null) {
+            final long val = Timestamps.parseOffset(offset);
+            if (val == Numbers.LONG_NaN) {
+                // bad value for offset
+                Misc.free(base);
```

Comment:
This is an excellent spot, thanks Alex. Double close essentially, it only works because underlying cursor handles double-close gracefully. But not all of them do.